File: src/ios/NotificarePushPluginEventBroker.swift
import NotificareKit

class NotificarePushPluginEventBroker {

    typealias Consumer = (_ event: Event) -> Void

    private static var eventQueue = [Event]()
    private static var consumer: Consumer?
    private static var canEmitEvents = false

    static func startListening(settings: [AnyHashable : Any]?, _ consumer: @escaping Consumer) {
        let holdEventsUntilReady = settings?["re.notifica.cordova.hold_events_until_ready"] as? String == "true"

        self.consumer = consumer
        canEmitEvents = !holdEventsUntilReady || Notificare.shared.isReady

        if (!canEmitEvents) {
            NotificationCenter.default.addObserver(self, selector: #selector(self.onNotificareReady), name: Notification.Name("re.notifica.cordova.on_ready"), object: nil)
            return
        }

        processQueue()
    }

    static func dispatchEvent(name: String, payload: Any?) {
        let event = Event(name: name, payload: payload)

        if let consumer = consumer, canEmitEvents {
            consumer(event)
            return
        }

        eventQueue.append(event)
    }

    static private func processQueue() {
        guard let consumer = consumer else {
            NotificareLogger.debug("Cannot process event queue without a consumer.")
            return
        }

        guard !eventQueue.isEmpty else { return }

        NotificareLogger.debug("Processing event queue with ${eventQueue.size} items.")
        eventQueue.forEach { consumer($0) }
        eventQueue.removeAll()
    }

    @objc static func onNotificareReady() {
        NotificationCenter.default.removeObserver(self)

        canEmitEvents = true
        processQueue()
    }

    private init() {}

    struct Event {
        let name: String
        let payload: Any?
    }
}

File: src/android/NotificarePushPluginEventBroker.kt
package re.notifica.push.cordova

import android.os.Handler
import android.os.Looper
import org.apache.cordova.CordovaPreferences
import re.notifica.Notificare
import re.notifica.internal.NotificareLogger
import re.notifica.models.NotificareApplication

internal object NotificarePushPluginEventBroker : Notificare.Listener {

    private val eventQueue = mutableListOf<Event>()
    private var consumer: Consumer? = null
    private var canEmitEvents = false

    fun dispatchEvent(name: String, payload: Any?) {
        val event = Event(name, payload)

        val consumer = consumer
        if (consumer != null && canEmitEvents) {
            consumer.onEvent(event)
            return
        }

        eventQueue.add(event)
    }

    fun setup(preferences: CordovaPreferences, consumer: Consumer) {
        val holdEventsUntilReady = preferences.getBoolean("re.notifica.cordova.hold_events_until_ready", false)

        this.consumer = consumer
        canEmitEvents = !holdEventsUntilReady || Notificare.isReady

        if (!canEmitEvents) {
            Notificare.addListener(this)
            return
        }

        processQueue()
    }

    private fun processQueue() {
        val consumer = consumer ?: run {
            NotificareLogger.debug("Cannot process event queue without a consumer.")
            return
        }

        if (eventQueue.isEmpty()) return

        NotificareLogger.debug("Processing event queue with ${eventQueue.size} items.")
        eventQueue.forEach { consumer.onEvent(it) }
        eventQueue.clear()
    }

    override fun onReady(application: NotificareApplication) {
        // Run the listener removal on the next loop to prevent modifying the collection
        // while iterating on it, causing an exception.
        Handler(Looper.getMainLooper()).post {
            Notificare.removeListener(this)
        }

        canEmitEvents = true
        processQueue()
    }


    data class Event(
        val name: String,
        val payload: Any?,
    )

    interface Consumer {
        fun onEvent(event: Event)
    }
}
